src/lib/zkstring.ts

```diff
@@ -0,0 +1,148 @@
+import { Bool, Circuit, Field, Poseidon } from '../snarky';
+import { CircuitValue, prop } from './circuit_value';
+
+export class Character extends CircuitValue {
+  @prop value: Field;
+
+  constructor(value: Field) {
+    super();
+
+    if (Character.validateInput(value)) {
+      this.value = value;
+    } else {
+      throw Error('Character initialization did not pass validation');
+    }
+  }
+
+  equals(char: Character): Bool {
+    return this.value.equals(char.value);
+  }
+
+  toField(): Field {
+    return this.value;
+  }
+
+  toString(): string {
+    const charCode = Number(this.value.toString());
+    return String.fromCharCode(charCode);
+  }
+
+  static fromString(str: string) {
+    const char = Field(str.charCodeAt(0));
+    return new Character(char);
+  }
+
+  /* 
+  TODO: Add support for more character sets
+  ASCII reference: https://en.wikipedia.org/wiki/ASCII#Printable_characters
+  */
+  static validateInput(value: Field): Boolean {
+    if (Circuit.inProver() || Circuit.inCheckedComputation()) {
+      const ret = Bool.and(
+        value.gte(Field(32)),
```

Comment:
I would go for a full bit range, 0 to 255, because that can be implemented more efficiently.
Use rangeCheckHelper(8), like in UInt32

Reply:
Hmm, the range here was intentionally chosen from: https://en.wikipedia.org/wiki/ASCII#Printable_characters

I didn't want to get too deep into the implementation of any one character set, but this range is specifically printable ascii characters.  In the future, people will probably want much richer characters which will require even more rules.  Is this less efficient version going to be like unacceptably slow?  If so, it seems to be a soft blocker on developing this string concept any further.

Reply:
Gotcha. It's not _unacceptably_ slow. However, a full bit range can be made very fast with lookup tables (coming soon), and detailed ranges will probably never be as fast.

In the current implementation, with `String.fromCharCode` as the mapping between numbers and chars, I think the range 0 to 255 is just fine. It can be converted back and forth to a string unambiguously. So I don't see the argument for introducing a performance penalty, just so it belongs to a set of "printable" characters.